### PR TITLE
[SPIR-V 1.6] True Label and False Label of an OpBranchConditional must not be the same

### PR DESCRIPTION
**Pull request by vmaksimo**

### CODE REVIEW

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 1]
```diff
@@ -859,6 +859,8 @@ class SPIRVBranchConditional : public SPIRVInstruction {
            getCondition()->getType()->isTypeBool());
     assert(getTrueLabel()->isForward() || getTrueLabel()->isLabel());
     assert(getFalseLabel()->isForward() || getFalseLabel()->isLabel());
+    if (Module->isAllowedToUseVersion(VersionNumber::SPIRV_1_6))
```
[the remark this answers is not in the document]
Comment (vmaksimo) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:862]: @MrSidims Actually we didn't decide anything about error mechanism: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/2482#discussion_r1561141986
I'm not sure again if it's related to the PR, but it seems we should rework all the `validate()` methods to get rid of assertions. Will this be one day replaced with the validator from the SPIRV-Tools?

[the remark this answers is not in the document]
Comment (LU-JOHN) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:862]: Isn't this too dangerous?   What if someone had input LLVM that previously worked, but it starts failing with this assert?

[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:862]: @LU-JOHN would you mind to provide an example of such module?

[the remark this answers is not in the document]
Comment (LU-JOHN) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:862]: This code is silly, but valid:

> define spir_func void @func_export(i16 %a) {
> entry:
>   %cmp = icmp eq i16 %a, 0
>   br i1 %cmp, label %same, label %same
> 
> same:
>   ret void
> }

[the remark this answers is not in the document]
Comment (asudarsa) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:862]: I wonder why this restriction was added. What positive impact does it have?

On a more relevant note, I think this PR might need a test. Thanks

[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:862]: > This code is silly, but valid:

And how can we generate such module to break existing flows? But may be we indeed shall not add an assertion, instead we should just skip such branch instructions during translation, but I'd like to see C examples. I've attempted to get one with goto, but failed. We can't get such BB as a loop's latch BB, right?

> What positive impact does it have?

If I'm not mistaken, the idea is to ensure maximum convergence in generated SPIR-V modules, for this you need to have as minimum re-convergence points as possible (there is more work for this, but it's not applicable to this very change). In the example provided by John the branch instruction is useless and can be removed supporting minimum re-convergence goal.

[the remark this answers is not in the document]
Comment (LU-JOHN) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:862]: I couldn't make an example from C either.

[the remark this answers is not in the document]
Comment (vmaksimo) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:862]: @MrSidims @LU-JOHN @asudarsa To sum it up, would you suggest not to have this change at all, or is it better to commit as is without the test to be more specification conformant?

**Review by MrSidims — APPROVED** [review]

**Pull request merged**